Replace PNG images with SVG in Tango art provider

TcT2k opened the pull request:
* Replace all PNG images from the tango icon set with their SVG versions
* Add refresh and stop art ids and use them in webview sample

All SVG are size optimized with [svgo](/svg/svgo) before converted to header by `bin2c.py`

TcT2k committed:
Use SVG instead of PNGs in tango art provider

Replace embedded tango PNG data with SVG data of the same icons.

TcT2k committed:
Add tango art for refresh and stop

Use the added art IDs in webview sample

TcT2k committed:
Add additional wxBitmapBundle::FromSVG() overload

This overload can work on non 0 terminated unsigned char arrays.
Especially useful for headers created by bin2c.py

TcT2k commented:
Not really sure how the single mac configuration build failure is related to my changes. But it at least has something to do with `wxArtProvider` @vadz maybe you have an idea how this could be related?

PBfordev commented:
Thanks @TcT2k, you are fast.

I thought we would keep the PNG implementation and switch between the SVGs and PNGs only at build-time, depending on whether SVG rasterization is available? But perhaps there are no platforms using `wxArtTangoProvider` without SVG support?

I can't say I understand the `wxArtProvider`/`wxBitmapBundleImplArt` code, for example, so I will not comment on the actual implementation.

However, after running the artprov sample and using Resource Browser there under MSVS, I can see there are many memory leaks reported after quitting the sample:
<details>
<summary>Leak report</summary>
Detected memory leaks!
Dumping objects ->
{32708} normal block at 0x008BC1A0, 32 bytes long.
 Data: <w x A R T _ S T > 77 00 78 00 41 00 52 00 54 00 5F 00 53 00 54 00 
{32707} normal block at 0x008E4EE0, 8 bytes long.
 Data: <p       > 70 BF 8B 00 00 00 00 00 
{32704} normal block at 0x008BBF70, 36 bytes long.
 Data: < N              > E0 4E 8E 00 A0 C1 8B 00 CD CD CD CD CD CD CD CD 
{32528} normal block at 0x008BC330, 32 bytes long.
 Data: <w x A R T _ R E > 77 00 78 00 41 00 52 00 54 00 5F 00 52 00 45 00 
{32527} normal block at 0x008E4AF0, 8 bytes long.
 Data: <        > D0 C3 8B 00 00 00 00 00 
{32524} normal block at 0x008BC3D0, 36 bytes long.
 Data: < J  0           > F0 4A 8E 00 30 C3 8B 00 CD CD CD CD CD CD CD CD 
{32330} normal block at 0x008BBF20, 32 bytes long.
 Data: <w x A R T _ R E > 77 00 78 00 41 00 52 00 54 00 5F 00 52 00 45 00 
{32329} normal block at 0x008E49D8, 8 bytes long.
 Data: <0       > 30 BE 8B 00 00 00 00 00 
{32326} normal block at 0x008BBE30, 36 bytes long.
 Data: < I              > D8 49 8E 00 20 BF 8B 00 CD CD CD CD CD CD CD CD 
{32238} normal block at 0x008BC150, 32 bytes long.
 Data: <w x A R T _ C D > 77 00 78 00 41 00 52 00 54 00 5F 00 43 00 44 00 
{32237} normal block at 0x008E4BD0, 8 bytes long.
 Data: <p       > 70 C4 8B 00 00 00 00 00 
{32234} normal block at 0x008BC470, 36 bytes long.
 Data: < K  P           > D0 4B 8E 00 50 C1 8B 00 CD CD CD CD CD CD CD CD 
{32149} normal block at 0x008BC2E0, 32 bytes long.
 Data: <w x A R T _ F L > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 4C 00 
{32148} normal block at 0x008E4770, 8 bytes long.
 Data: <`       > 60 C0 8B 00 00 00 00 00 
{32145} normal block at 0x008BC060, 36 bytes long.
 Data: <pG              > 70 47 8E 00 E0 C2 8B 00 CD CD CD CD CD CD CD CD 
{32063} normal block at 0x008BC4C0, 32 bytes long.
 Data: <w x A R T _ H A > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 41 00 
{32062} normal block at 0x008E4658, 8 bytes long.
 Data: <        > E0 BD 8B 00 00 00 00 00 
{32059} normal block at 0x008BBDE0, 36 bytes long.
 Data: <XF              > 58 46 8E 00 C0 C4 8B 00 CD CD CD CD CD CD CD CD 
{31978} normal block at 0x008C76B0, 48 bytes long.
 Data: <w x A R T _ F U > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 55 00 
{31977} normal block at 0x008E4690, 8 bytes long.
 Data: <        > A0 BC 8B 00 00 00 00 00 
{31974} normal block at 0x008BBCA0, 36 bytes long.
 Data: < F   v          > 90 46 8E 00 B0 76 8C 00 CD CD CD CD CD CD CD CD 
{31772} normal block at 0x008BF8F0, 32 bytes long.
 Data: <w x A R T _ E D > 77 00 78 00 41 00 52 00 54 00 5F 00 45 00 44 00 
{31771} normal block at 0x008E4B28, 8 bytes long.
 Data: <p       > 70 FB 8B 00 00 00 00 00 
{31768} normal block at 0x008BFB70, 36 bytes long.
 Data: <(K              > 28 4B 8E 00 F0 F8 8B 00 CD CD CD CD CD CD CD CD 
{31015} normal block at 0x008C7110, 48 bytes long.
 Data: <w x A R T _ F I > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 49 00 
{31014} normal block at 0x008E4C40, 8 bytes long.
 Data: <        > A0 F8 8B 00 00 00 00 00 
{31011} normal block at 0x008BF8A0, 36 bytes long.
 Data: <@L   q          > 40 4C 8E 00 10 71 8C 00 CD CD CD CD CD CD CD CD 
{30660} normal block at 0x008BE900, 32 bytes long.
 Data: <w x A R T _ F I > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 49 00 
{30659} normal block at 0x008E4818, 8 bytes long.
 Data: <        > 10 E8 8B 00 00 00 00 00 
{30656} normal block at 0x008BE810, 36 bytes long.
 Data: < H              > 18 48 8E 00 00 E9 8B 00 CD CD CD CD CD CD CD CD 
{30577} normal block at 0x008BE590, 32 bytes long.
 Data: <w x A R T _ Q U > 77 00 78 00 41 00 52 00 54 00 5F 00 51 00 55 00 
{30576} normal block at 0x008E49A0, 8 bytes long.
 Data: <0       > 30 EB 8B 00 00 00 00 00 
{30573} normal block at 0x008BEB30, 36 bytes long.
 Data: < I              > A0 49 8E 00 90 E5 8B 00 CD CD CD CD CD CD CD CD 
{29443} normal block at 0x008BE720, 32 bytes long.
 Data: <w x A R T _ C L > 77 00 78 00 41 00 52 00 54 00 5F 00 43 00 4C 00 
{29442} normal block at 0x008C4FD8, 8 bytes long.
 Data: <P       > 50 E9 8B 00 00 00 00 00 
{29439} normal block at 0x008BE950, 36 bytes long.
 Data: < O              > D8 4F 8C 00 20 E7 8B 00 CD CD CD CD CD CD CD CD 
{28953} normal block at 0x008BEAE0, 32 bytes long.
 Data: <w x A R T _ M I > 77 00 78 00 41 00 52 00 54 00 5F 00 4D 00 49 00 
{28952} normal block at 0x008C5010, 8 bytes long.
 Data: <@       > 40 E5 8B 00 00 00 00 00 
{28949} normal block at 0x008BE540, 36 bytes long.
 Data: < P              > 10 50 8C 00 E0 EA 8B 00 CD CD CD CD CD CD CD CD 
{28772} normal block at 0x008BE4A0, 32 bytes long.
 Data: <w x A R T _ P L > 77 00 78 00 41 00 52 00 54 00 5F 00 50 00 4C 00 
{28771} normal block at 0x008C4F68, 8 bytes long.
 Data: <        > F0 E4 8B 00 00 00 00 00 
{28768} normal block at 0x008BE4F0, 36 bytes long.
 Data: <hO              > 68 4F 8C 00 A0 E4 8B 00 CD CD CD CD CD CD CD CD 
{28585} normal block at 0x008BE8B0, 32 bytes long.
 Data: <w x A R T _ R E > 77 00 78 00 41 00 52 00 54 00 5F 00 52 00 45 00 
{28584} normal block at 0x008C5550, 8 bytes long.
 Data: <P       > 50 EE 8B 00 00 00 00 00 
{28581} normal block at 0x008BEE50, 36 bytes long.
 Data: <PU              > 50 55 8C 00 B0 E8 8B 00 CD CD CD CD CD CD CD CD 
{28420} normal block at 0x008BEE00, 32 bytes long.
 Data: <w x A R T _ U N > 77 00 78 00 41 00 52 00 54 00 5F 00 55 00 4E 00 
{28419} normal block at 0x008C5128, 8 bytes long.
 Data: <@       > 40 EA 8B 00 00 00 00 00 
{28416} normal block at 0x008BEA40, 36 bytes long.
 Data: <(Q              > 28 51 8C 00 00 EE 8B 00 CD CD CD CD CD CD CD CD 
{28255} normal block at 0x008BEA90, 32 bytes long.
 Data: <w x A R T _ P A > 77 00 78 00 41 00 52 00 54 00 5F 00 50 00 41 00 
{28254} normal block at 0x008C5240, 8 bytes long.
 Data: <        > E0 E5 8B 00 00 00 00 00 
{28251} normal block at 0x008BE5E0, 36 bytes long.
 Data: <@R              > 40 52 8C 00 90 EA 8B 00 CD CD CD CD CD CD CD CD 
{27996} normal block at 0x008BED10, 32 bytes long.
 Data: <w x A R T _ C U > 77 00 78 00 41 00 52 00 54 00 5F 00 43 00 55 00 
{27995} normal block at 0x008C5358, 8 bytes long.
 Data: <        > 80 E6 8B 00 00 00 00 00 
{27992} normal block at 0x008BE680, 36 bytes long.
 Data: <XS              > 58 53 8C 00 10 ED 8B 00 CD CD CD CD CD CD CD CD 
{27757} normal block at 0x008BE6D0, 32 bytes long.
 Data: <w x A R T _ C O > 77 00 78 00 41 00 52 00 54 00 5F 00 43 00 4F 00 
{27756} normal block at 0x008C5080, 8 bytes long.
 Data: <0       > 30 E6 8B 00 00 00 00 00 
{27753} normal block at 0x008BE630, 36 bytes long.
 Data: < P              > 80 50 8C 00 D0 E6 8B 00 CD CD CD CD CD CD CD CD 
{27553} normal block at 0x008BE7C0, 32 bytes long.
 Data: <w x A R T _ D E > 77 00 78 00 41 00 52 00 54 00 5F 00 44 00 45 00 
{27552} normal block at 0x008C6E78, 8 bytes long.
 Data: <p       > 70 E7 8B 00 00 00 00 00 
{27549} normal block at 0x008BE770, 36 bytes long.
 Data: <xn              > 78 6E 8C 00 C0 E7 8B 00 CD CD CD CD CD CD CD CD 
{27473} normal block at 0x008C6F90, 48 bytes long.
 Data: <w x A R T _ F I > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 49 00 
{27472} normal block at 0x008C6BD8, 8 bytes long.
 Data: <p       > 70 EC 8B 00 00 00 00 00 
{27469} normal block at 0x008BEC70, 36 bytes long.
 Data: < k   o          > D8 6B 8C 00 90 6F 8C 00 CD CD CD CD CD CD CD CD 
{27166} normal block at 0x008BDF00, 32 bytes long.
 Data: <w x A R T _ F I > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 49 00 
{27165} normal block at 0x008C6C48, 8 bytes long.
 Data: <P       > 50 E4 8B 00 00 00 00 00 
{27162} normal block at 0x008BE450, 36 bytes long.
 Data: <Hl              > 48 6C 8C 00 00 DF 8B 00 CD CD CD CD CD CD CD CD 
{26872} normal block at 0x008BE220, 32 bytes long.
 Data: <w x A R T _ F I > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 49 00 
{26871} normal block at 0x008C6D60, 8 bytes long.
 Data: <        > B0 E3 8B 00 00 00 00 00 
{26868} normal block at 0x008BE3B0, 36 bytes long.
 Data: <`m              > 60 6D 8C 00 20 E2 8B 00 CD CD CD CD CD CD CD CD 
{26583} normal block at 0x008BDF50, 32 bytes long.
 Data: <w x A R T _ N E > 77 00 78 00 41 00 52 00 54 00 5F 00 4E 00 45 00 
{26582} normal block at 0x008C6C10, 8 bytes long.
 Data: <        > D0 E1 8B 00 00 00 00 00 
{26579} normal block at 0x008BE1D0, 36 bytes long.
 Data: < l  P           > 10 6C 8C 00 50 DF 8B 00 CD CD CD CD CD CD CD CD 
{26372} normal block at 0x008C6F30, 48 bytes long.
 Data: <w x A R T _ M I > 77 00 78 00 41 00 52 00 54 00 5F 00 4D 00 49 00 
{26371} normal block at 0x008C66D0, 8 bytes long.
 Data: <`       > 60 DE 8B 00 00 00 00 00 
{26368} normal block at 0x008BDE60, 36 bytes long.
 Data: < f  0o          > D0 66 8C 00 30 6F 8C 00 CD CD CD CD CD CD CD CD 
{26189} normal block at 0x008C7410, 48 bytes long.
 Data: <w x A R T _ C R > 77 00 78 00 41 00 52 00 54 00 5F 00 43 00 52 00 
{26188} normal block at 0x008C6820, 8 bytes long.
 Data: <`       > 60 D4 8B 00 00 00 00 00 
{26185} normal block at 0x008BD460, 36 bytes long.
 Data: < h   t          > 20 68 8C 00 10 74 8C 00 CD CD CD CD CD CD CD CD 
{25761} normal block at 0x008BD500, 32 bytes long.
 Data: <w x A R T _ T I > 77 00 78 00 41 00 52 00 54 00 5F 00 54 00 49 00 
{25760} normal block at 0x008C6468, 8 bytes long.
 Data: <        > 10 D4 8B 00 00 00 00 00 
{25757} normal block at 0x008BD410, 36 bytes long.
 Data: <hd              > 68 64 8C 00 00 D5 8B 00 CD CD CD CD CD CD CD CD 
{25334} normal block at 0x008C7AD0, 48 bytes long.
 Data: <w x A R T _ N O > 77 00 78 00 41 00 52 00 54 00 5F 00 4E 00 4F 00 
{25333} normal block at 0x008C64A0, 8 bytes long.
 Data: <        > C0 D3 8B 00 00 00 00 00 
{25330} normal block at 0x008BD3C0, 36 bytes long.
 Data: < d   z          > A0 64 8C 00 D0 7A 8C 00 CD CD CD CD CD CD CD CD 
{25237} normal block at 0x008C7950, 48 bytes long.
 Data: <w x A R T _ E X > 77 00 78 00 41 00 52 00 54 00 5F 00 45 00 58 00 
{25236} normal block at 0x008C69A8, 8 bytes long.
 Data: <        > A0 D0 8B 00 00 00 00 00 
{25233} normal block at 0x008BD0A0, 36 bytes long.
 Data: < i  Py          > A8 69 8C 00 50 79 8C 00 CD CD CD CD CD CD CD CD 
{25143} normal block at 0x008BD690, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{25142} normal block at 0x008C6740, 8 bytes long.
 Data: <        > E0 D6 8B 00 00 00 00 00 
{25139} normal block at 0x008BD6E0, 36 bytes long.
 Data: <@g              > 40 67 8C 00 90 D6 8B 00 CD CD CD CD CD CD CD CD 
{24269} normal block at 0x0087DF60, 48 bytes long.
 Data: <w x A R T _ F O > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 4F 00 
{24268} normal block at 0x008C68C8, 8 bytes long.
 Data: <        > D0 D2 8B 00 00 00 00 00 
{24265} normal block at 0x008BD2D0, 36 bytes long.
 Data: < h  `           > C8 68 8C 00 60 DF 87 00 CD CD CD CD CD CD CD CD 
{24190} normal block at 0x008BD5F0, 32 bytes long.
 Data: <w x A R T _ F O > 77 00 78 00 41 00 52 00 54 00 5F 00 46 00 4F 00 
{24189} normal block at 0x008C6938, 8 bytes long.
 Data: <0       > 30 D2 8B 00 00 00 00 00 
{24186} normal block at 0x008BD230, 36 bytes long.
 Data: <8i              > 38 69 8C 00 F0 D5 8B 00 CD CD CD CD CD CD CD CD 
{24114} normal block at 0x008BD640, 32 bytes long.
 Data: <w x A R T _ N E > 77 00 78 00 41 00 52 00 54 00 5F 00 4E 00 45 00 
{24113} normal block at 0x008C6430, 8 bytes long.
 Data: <        > A0 D5 8B 00 00 00 00 00 
{24110} normal block at 0x008BD5A0, 36 bytes long.
 Data: <0d  @           > 30 64 8C 00 40 D6 8B 00 CD CD CD CD CD CD CD CD 
{23867} normal block at 0x008BD1E0, 32 bytes long.
 Data: <w x A R T _ L I > 77 00 78 00 41 00 52 00 54 00 5F 00 4C 00 49 00 
{23866} normal block at 0x008C6388, 8 bytes long.
 Data: <p       > 70 D8 8B 00 00 00 00 00 
{23863} normal block at 0x008BD870, 36 bytes long.
 Data: < c              > 88 63 8C 00 E0 D1 8B 00 CD CD CD CD CD CD CD CD 
{23188} normal block at 0x008BBB28, 48 bytes long.
 Data: <w x A R T _ R E > 77 00 78 00 41 00 52 00 54 00 5F 00 52 00 45 00 
{23187} normal block at 0x008C6318, 8 bytes long.
 Data: <p       > 70 D3 8B 00 00 00 00 00 
{23184} normal block at 0x008BD370, 36 bytes long.
 Data: < c  (           > 18 63 8C 00 28 BB 8B 00 CD CD CD CD CD CD CD CD 
{22516} normal block at 0x008BD550, 32 bytes long.
 Data: <w x A R T _ T I > 77 00 78 00 41 00 52 00 54 00 5F 00 54 00 49 00 
{22515} normal block at 0x008C5FD0, 8 bytes long.
 Data: <`       > 60 D9 8B 00 00 00 00 00 
{22512} normal block at 0x008BD960, 36 bytes long.
 Data: < _  P           > D0 5F 8C 00 50 D5 8B 00 CD CD CD CD CD CD CD CD 
{20226} normal block at 0x008BDA50, 32 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{20225} normal block at 0x008C60E8, 8 bytes long.
 Data: <        > F0 D0 8B 00 00 00 00 00 
{20222} normal block at 0x008BD0F0, 36 bytes long.
 Data: < `  P           > E8 60 8C 00 50 DA 8B 00 CD CD CD CD CD CD CD CD 
{20151} normal block at 0x008BD780, 32 bytes long.
 Data: <w x A R T _ P R > 77 00 78 00 41 00 52 00 54 00 5F 00 50 00 52 00 
{20150} normal block at 0x008C5D68, 8 bytes long.
 Data: <0       > 30 D7 8B 00 00 00 00 00 
{20147} normal block at 0x008BD730, 36 bytes long.
 Data: <h]              > 68 5D 8C 00 80 D7 8B 00 CD CD CD CD CD CD CD CD 
{20058} normal block at 0x008BD4B0, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{20057} normal block at 0x008C6200, 8 bytes long.
 Data: <        > 90 D1 8B 00 00 00 00 00 
{20054} normal block at 0x008BD190, 36 bytes long.
 Data: < b              > 00 62 8C 00 B0 D4 8B 00 CD CD CD CD CD CD CD CD 
{19896} normal block at 0x008BB168, 48 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{19895} normal block at 0x008C6008, 8 bytes long.
 Data: <        > 10 D9 8B 00 00 00 00 00 
{19892} normal block at 0x008BD910, 36 bytes long.
 Data: < `  h           > 08 60 8C 00 68 B1 8B 00 CD CD CD CD CD CD CD CD 
{19734} normal block at 0x008BDA00, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{19733} normal block at 0x008C5E10, 8 bytes long.
 Data: <        > C0 D8 8B 00 00 00 00 00 
{19730} normal block at 0x008BD8C0, 36 bytes long.
 Data: < ^              > 10 5E 8C 00 00 DA 8B 00 CD CD CD CD CD CD CD CD 
{19480} normal block at 0x008BAF28, 48 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{19479} normal block at 0x008C6040, 8 bytes long.
 Data: <@       > 40 D1 8B 00 00 00 00 00 
{19476} normal block at 0x008BD140, 36 bytes long.
 Data: <@`  (           > 40 60 8C 00 28 AF 8B 00 CD CD CD CD CD CD CD CD 
{18807} normal block at 0x008BD280, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{18806} normal block at 0x008C6350, 8 bytes long.
 Data: <        > B0 D9 8B 00 00 00 00 00 
{18803} normal block at 0x008BD9B0, 36 bytes long.
 Data: <Pc              > 50 63 8C 00 80 D2 8B 00 CD CD CD CD CD CD CD CD 
{18651} normal block at 0x008BD820, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{18650} normal block at 0x008C6190, 8 bytes long.
 Data: <        > D0 D7 8B 00 00 00 00 00 
{18647} normal block at 0x008BD7D0, 36 bytes long.
 Data: < a              > 90 61 8C 00 20 D8 8B 00 CD CD CD CD CD CD CD CD 
{18494} normal block at 0x008BB408, 48 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{18493} normal block at 0x008A08A0, 8 bytes long.
 Data: <        > 20 D3 8B 00 00 00 00 00 
{18490} normal block at 0x008BD320, 36 bytes long.
 Data: <                > A0 08 8A 00 08 B4 8B 00 CD CD CD CD CD CD CD CD 
{18337} normal block at 0x008BCCE0, 32 bytes long.
 Data: <w x A R T _ G O > 77 00 78 00 41 00 52 00 54 00 5F 00 47 00 4F 00 
{18336} normal block at 0x008A0788, 8 bytes long.
 Data: <        > 20 CE 8B 00 00 00 00 00 
{18333} normal block at 0x008BCE20, 36 bytes long.
 Data: <                > 88 07 8A 00 E0 CC 8B 00 CD CD CD CD CD CD CD CD 
{18126} normal block at 0x008BCBF0, 32 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{18125} normal block at 0x008A0638, 8 bytes long.
 Data: <        > 00 D0 8B 00 00 00 00 00 
{18122} normal block at 0x008BD000, 36 bytes long.
 Data: <8               > 38 06 8A 00 F0 CB 8B 00 CD CD CD CD CD CD CD CD 
{17452} normal block at 0x008BB528, 48 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{17451} normal block at 0x008A0718, 8 bytes long.
 Data: <        > E0 C7 8B 00 00 00 00 00 
{17448} normal block at 0x008BC7E0, 36 bytes long.
 Data: <    (           > 18 07 8A 00 28 B5 8B 00 CD CD CD CD CD CD CD CD 
{16769} normal block at 0x008BC880, 32 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{16768} normal block at 0x008A0C90, 8 bytes long.
 Data: <P       > 50 CB 8B 00 00 00 00 00 
{16765} normal block at 0x008BCB50, 36 bytes long.
 Data: <                > 90 0C 8A 00 80 C8 8B 00 CD CD CD CD CD CD CD CD 
{16087} normal block at 0x008BB648, 48 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{16086} normal block at 0x008A0600, 8 bytes long.
 Data: <        > D0 CD 8B 00 00 00 00 00 
{16083} normal block at 0x008BCDD0, 36 bytes long.
 Data: <    H           > 00 06 8A 00 48 B6 8B 00 CD CD CD CD CD CD CD CD 
{15444} normal block at 0x008BAFE8, 48 bytes long.
 Data: <w x A R T _ H E > 77 00 78 00 41 00 52 00 54 00 5F 00 48 00 45 00 
{15443} normal block at 0x008A09F0, 8 bytes long.
 Data: <        > 00 CB 8B 00 00 00 00 00 
{15440} normal block at 0x008BCB00, 36 bytes long.
 Data: <                > F0 09 8A 00 E8 AF 8B 00 CD CD CD CD CD CD CD CD 
{14779} normal block at 0x008BB1C8, 48 bytes long.
 Data: <w x A R T _ D E > 77 00 78 00 41 00 52 00 54 00 5F 00 44 00 45 00 
{14778} normal block at 0x008A09B8, 8 bytes long.
 Data: <        > F0 C6 8B 00 00 00 00 00 
{14775} normal block at 0x008BC6F0, 36 bytes long.
 Data: <                > B8 09 8A 00 C8 B1 8B 00 CD CD CD CD CD CD CD CD 
{14117} normal block at 0x008BB3A8, 48 bytes long.
 Data: <w x A R T _ A D > 77 00 78 00 41 00 52 00 54 00 5F 00 41 00 44 00 
{14116} normal block at 0x008A1E10, 8 bytes long.
 Data: <        > B0 CF 8B 00 00 00 00 00 
{14113} normal block at 0x008BCFB0, 36 bytes long.
 Data: <                > 10 1E 8A 00 A8 B3 8B 00 CD CD CD CD CD CD CD CD 
{13301} normal block at 0x0087DE40, 48 bytes long.
 Data: <w x A R T _ I N > 77 00 78 00 41 00 52 00 54 00 5F 00 49 00 4E 00 
{13300} normal block at 0x008A1E48, 8 bytes long.
 Data: < :      > 80 3A 86 00 00 00 00 00 
{13297} normal block at 0x00863A80, 36 bytes long.
 Data: <H   @           > 48 1E 8A 00 40 DE 87 00 CD CD CD CD CD CD CD CD 
{13228} normal block at 0x00863620, 32 bytes long.
 Data: <w x A R T _ W A > 77 00 78 00 41 00 52 00 54 00 5F 00 57 00 41 00 
{13227} normal block at 0x008A1B00, 8 bytes long.
 Data: <@>      > 40 3E 86 00 00 00 00 00 
{13224} normal block at 0x00863E40, 36 bytes long.
 Data: <     6          > 00 1B 8A 00 20 36 86 00 CD CD CD CD CD CD CD CD 
{13158} normal block at 0x00863990, 32 bytes long.
 Data: <w x A R T _ Q U > 77 00 78 00 41 00 52 00 54 00 5F 00 51 00 55 00 
{13157} normal block at 0x008A1B70, 8 bytes long.
 Data: <@9      > 40 39 86 00 00 00 00 00 
{13154} normal block at 0x00863940, 36 bytes long.
 Data: <p    9          > 70 1B 8A 00 90 39 86 00 CD CD CD CD CD CD CD CD 
{13088} normal block at 0x00863850, 32 bytes long.
 Data: <w x A R T _ E R > 77 00 78 00 41 00 52 00 54 00 5F 00 45 00 52 00 
{13087} normal block at 0x008A1630, 8 bytes long.
 Data: < ?      > 80 3F 86 00 00 00 00 00 
{13084} normal block at 0x00863F80, 36 bytes long.
 Data: <0   P8          > 30 16 8A 00 50 38 86 00 CD CD CD CD CD CD CD CD 
Object dump complete.
</details>

BTW, the bitmaps in the webview sample now look rather small. It is not only the wxlogo which has large margins, but the other bitmaps look small compared to the ones in the toolbar sample too (DPI at 125%, the toolbar sample does not use wxlogo from `wxArtProvider`):
![webview-sample-bitmaps-tango](https://user-images.githubusercontent.com/12495521/151537941-d490baa4-64a7-42bd-b697-fb2c1de4ba72.png)

TcT2k commented:
> I thought we would keep the PNG implementation and switch between the SVGs and PNGs only at build-time, depending on whether SVG rasterization is available? But perhaps there are no platforms using `wxArtTangoProvider` without SVG support?

I don't think it's worth keeping the PNGs (and it makes maintenance much easier).

> 
> I can't say I understand the `wxArtProvider`/`wxBitmapBundleImplArt` code, so I will not comment on the actual implementation.
> 

> BTW, the bitmaps in the webview sample now look rather small. It is not only the wxlogo which has large margins, but the other bitmaps look small compared to the ones in the toolbar sample too (DPI at 125%, the toolbar sample does not use wxlogo from `wxArtProvider`): ![webview-sample-bitmaps-tango](https://user-images.githubusercontent.com/12495521/151537941-d490baa4-64a7-42bd-b697-fb2c1de4ba72.png)
> 
Seemed a little small to me, but couldn't yet figure out why. I'll look into whats the difference with the toolbar sample.

> **EDIT** I think that both memory issues (leak on MSVS, overwrite MacOS) are there because `nsvgParse()` [requires 0-terminated string](https://github.com/wxWidgets/nanosvg/blob/ccdb1995134d340a93fb20e3a3d323ccb3838dd0/src/nanosvg.h#L168-L169), which the new `wxBitmapBundle::FromSVG()` overload using `wxCharBuffer::CreateNonOwned()` does not provide.

Good to know should be easy enough to add, will look into that. I was under the impression that the charbuffer would provide the terminating 0.

PBfordev commented:
>I was under the impression that the charbuffer would provide the terminating 0.

Sometimes it does (e.g., [`wxCharBuffer(size_t)` ctor](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/include/wx/buffer.h#L269)) but often/usually does not. I don't think it would be even possible with a non-owned buffer.

TBH, I do not understand how `wxBitmapBundle::FromSVG((const char*)wxlogo_svg_data, sizeDef);` [can work](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/src/common/artstd.cpp#L282).  [`wxlogo_svg_data`](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/art/wxlogo_svg.cpp#L742) is not 0-terminated, yet `wxStrlen()` is called on it here
https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/include/wx/buffer.h#L259-L260
To me it, it looks it looks like a bug and it works just  because there happens to be a 0 following `wxlogo_svg_data` in the memory segment, perhaps due to how the globals are stored - are we really relying on that?

Hopefully, I am missing something and it is not a bug, @vadz?

TcT2k commented:
> > I was under the impression that the charbuffer would provide the terminating 0.
> 
> Sometimes it does (e.g., [`wxCharBuffer(size_t)` ctor](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/include/wx/buffer.h#L269) used in [`wxBitmapBundle::FromSVGFile()`](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/src/generic/bmpsvg.cpp#L240)) but often/usually does not. I don't think it would be even possible with a non-owned buffer.
> 
> TBH, I do not understand how `wxBitmapBundle::FromSVG((const char*)wxlogo_svg_data, sizeDef);` [can work](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/src/common/artstd.cpp#L282). [`wxlogo_svg_data`](https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/art/wxlogo_svg.cpp#L742) is not 0-terminated, yet `wxStrlen()` is called on it here
> 
> https://github.com/wxWidgets/wxWidgets/blob/b8c457e356d9592a5ab08b8eeaf9cf4ecaf54f69/include/wx/buffer.h#L259-L260
> 

That's why I added the new `FromSVG()` overload, because there where some funky side effects when depending on `wxStrlen()` (overlayed SVGs and other intresting effects).

I've added a new commit which should ensure the trailing 0.

TcT2k commented:
Added a fix for toolbar size in the webview sample:

![image](https://user-images.githubusercontent.com/5075894/151609546-df00eeee-06a1-4cb9-bbb0-ba488b7b3b87.png)

Looks much better now. The logo seems to have a little to much padding I think.

TcT2k commented:
@PBfordev I've had a look at the memory leak (in the art provider sample) and that seems unrelated to my changes. It's the same with current master.

PBfordev commented:
> @PBfordev I've had a look at the memory leak (in the art provider sample) and that seems unrelated to my changes. It's the same with current master.

Actually, I have just tested and they are there in 3.1.5 already. Sorry for the noise. Unfortunately, I was unable to persuade MSVC heap debugger or Visual Leak Detector to point to the places where these allocations happen. :(

BTW, I wonder if we should not add the two new arts also as XPMs to fall back to?

vadz commented on src/common/artstd.cpp:
```diff
@@ -279,7 +279,7 @@ wxDefaultArtProvider::CreateBitmapBundle(const wxArtID& id,
         }
 
-        bb = wxBitmapBundle::FromSVG((const char*)wxlogo_svg_data, sizeDef);
+        bb = wxBitmapBundle::FromSVG(wxlogo_svg_data, sizeof(wxlogo_svg_data),sizeDef);
```
Nitpick of the day:
```suggestion
        bb = wxBitmapBundle::FromSVG(wxlogo_svg_data, sizeof(wxlogo_svg_data), sizeDef);
```

vadz commented on include/wx/bmpbndl.h:
```diff
@@ -76,6 +76,9 @@ class WXDLLIMPEXP_CORE wxBitmapBundle
     // This overload currently makes a copy of the data.
     static wxBitmapBundle FromSVG(const char* data, const wxSize& sizeDef);
 
+    // This overload can work on non 0 terminted data
```
```suggestion
    // This overload can work on non 0 terminated data
```

I wonder why this hasn't been caught by codespell...

vadz commented on include/wx/bmpbndl.h:
```diff
@@ -76,6 +76,9 @@ class WXDLLIMPEXP_CORE wxBitmapBundle
     static wxBitmapBundle FromSVG(const char* data, const wxSize& sizeDef);
 
+    // This overload can work on non 0 terminted data
+    static wxBitmapBundle FromSVG(const unsigned char* data, size_t len, const wxSize& sizeDef);
```
Would it make sense to use `wxByte` here for clarity? I.e.
```suggestion
    static wxBitmapBundle FromSVG(const wxByte* data, size_t len, const wxSize& sizeDef);
```

vadz commented on src/common/arttango.cpp:
```diff
@@ -42,22 +42,22 @@
 #include "../../art/tango/document_print.h"
-#include "../../art/tango/document_save.h"
 #include "../../art/tango/document_save_as.h"
+#include "../../art/tango/document_save.h"
```
Why change this? IMO `save` should come before `save_as` in the header order...

TcT2k replied:
I've just sorted the lines with Visual Studio Code.

vadz replied:
I'd like to undo it, just to avoid gratuitous differences with master. I can do it myself, of course, if you don't mind.

vadz commented on src/common/arttango.cpp:
```diff
@@ -103,34 +106,41 @@ class wxTangoArtProvider : public wxArtProvider
 
-wxBitmap
-wxTangoArtProvider::CreateBitmap(const wxArtID& id,
+wxBitmap wxTangoArtProvider::CreateBitmap(const wxArtID& id,
+    const wxArtClient& client,
+    const wxSize& size)
+{
+    wxBitmapBundle bb = CreateBitmapBundle(id, client, size);
+    if (bb.IsOk())
+        return bb.GetBitmap(size);
+    else
+        return wxNullBitmap;
+}
```
This should be really done in `wxArtProvider::GetBitmap()` itself, this is not at all specific to Tango provider.

TcT2k replied:
I agree but by simply putting it in the current `wxArtProvider` implementation would that create a potential recursion? as the default `CreateBitmapBundle()` implementation also makes a call to `CreateBitmap()`.

vadz replied:
You're right, I think I made a mistake with implementing `CreateBitmapBundle()` like this. I've tried to fix it in #22078 now, which should make overriding `CreateBitmap()` here unnecessary, so if you don't disagree, I'll merge the other PR first and then merge this one without this part.

vadz commented on src/generic/bmpsvg.cpp:
```diff
@@ -218,6 +218,16 @@ wxBitmapBundle wxBitmapBundle::FromSVG(const char* data, const wxSize& sizeDef)
+    copy.AppendData(data, len);
+    copy.AppendByte(0);
+
+    return FromSVG((char*)copy.GetData(), sizeDef);
```
I'd still use `wxCharBuffer` here, it looks like we could do it simply like this (but untested):
```suggestion
    wxCharBuffer copy(len);
    memcpy(copy.data(), data, len);

    return FromSVG(copy.data(), sizeDef);
```
and avoid any casts.

vadz reviewed (commented):
Thanks a lot for fixing the problem with non-NUL-terminated buffer and, of course, also thanks for adding the bitmaps!

This should be merged a.s.a.p. but if you can improve a couple of small things mentioned in the comments, it would be great (if not, I'll just do it myself when I can).

Thanks again and thanks to @PBfordev for testing/review!

vadz committed:
Improve fallback logic in wxArtProvider::GetBitmap{,Bundle}()

This improves the changes of f78db92462 (Avoid bitmap scaling in
wxArtProvider::GetBitmapBundle(), 2021-12-17) and still uses a custom
bundle to avoid scaling the bitmap if possible, but does it in
GetBitmapBundle() itself rather than CreateBitmapBundle().

This allows to also use CreateBitmapBundle() from GetBitmap(), as there
is no possibility of infinite recursion due to calling each of these
functions from the other one any more, and so allows defining art
providers overriding only CreateBitmapBundle() instead of having to
always override both it and CreateBitmap().

Also add a unit test, even if just a trivial one, for these functions,
to at least check that they don't crash.

vadz commented:
Thanks again, I'm going to merge it after #22078 if nobody sees any problems with it. I'll rebase the commits to squash the changes in the last one with the previous ones, which means that they won't be signed by you any longer, so please rebase them yourself if you'd prefer to keep them signed.

TcT2k commented:
> Thanks again, I'm going to merge it after #22078 if nobody sees any problems with it. 
Looks good to me, I've rebased this PR on your changes and removed `CreateBitmap()` from the tango art provider.

> I'll rebase the commits to squash the changes in the last one with the previous ones, which means that they won't be signed by you any longer, so please rebase them yourself if you'd prefer to keep them signed.

I've no problem with the commits no longer being signed in general, but I've squashed the commits here for now.

vadz commented:
Great, thanks again!
